src/pm/hydra/tools/bootstrap/external/slurm_query_node_list.c

```diff
@@ -9,9 +9,263 @@
+        gpattern[1] = gpattern[0] + gmatch[0][3].rm_so;
+
+        /* match new group-1 pattern: ([|,|^)(00-12 | 14)(]|,|$) */
+        while (*gpattern[1] && regexec(&gmatch_new[1], gpattern[1], MAX_GMATCH, gmatch[1], 0) == 0) {
+            /* bound group-1 for range/element matching: 00-12, -> 00-12\0 | 14] -> 14\0 */
+            tmp[1] = *(gpattern[1] + gmatch[1][0].rm_eo);
+            *(gpattern[1] + gmatch[1][0].rm_eo) = 0;
+
+            /* select second atom in group-1 */
+            rpattern = gpattern[1] + gmatch[1][2].rm_so;
+            epattern = rpattern;
+
+            if (regexec(&rmatch_new, rpattern, MAX_RMATCH, rmatch, 0) == 0) {
+                /* matched range: (00)-(10) */
+                sprintf(rbegin, "%.*s%c",
+                        (int)(rmatch[1].rm_eo-rmatch[1].rm_so), rpattern+rmatch[1].rm_so,
+                        '\0');
+                sprintf(rend, "%.*s%c",
+                        (int)(rmatch[2].rm_eo-rmatch[2].rm_so), rpattern+rmatch[2].rm_so,
+                        '\0');
+                begin = atoi(rbegin);
+                end = atoi(rend);
+
+                /* expand range and add nodes to global node list */
+                for (j = begin; j <= end; j++) {
+                    sprintf(hostname, "%s%.*d%c",
+                            basename,
+                            (int)(rmatch[1].rm_eo-rmatch[1].rm_so), j,
+                            '\0');
+                    status = HYDU_add_to_node_list(hostname, tasks_per_node[k++], &global_node_list);
+                    HYDU_ERR_POP(status, "unable to add to node list\n");
+                }
+            } else if (regexec(&ematch_new, epattern, MAX_EMATCH, ematch, 0) == 0) {
+                /* matched element: (14) */
+                sprintf(rbegin, "%.*s%c",
+                        (int)(ematch[1].rm_eo-ematch[1].rm_so), epattern+ematch[1].rm_so,
+                        '\0');
+                sprintf(hostname, "%s%s%c",
+                        basename,
+                        rbegin,
+                        '\0');
+                status = HYDU_add_to_node_list(hostname, tasks_per_node[k++], &global_node_list);
+                HYDU_ERR_POP(status, "unable to add to node list\n");
+            }
+
+            /* unbound group-1 and move to next group-1: 00-10\0 -> 00-10, | 14\0 -> 14] */
+            *(gpattern[1] + gmatch[1][0].rm_eo) = tmp[1];
+            gpattern[1] += gmatch[1][0].rm_eo;
+        }
+
+        /* unbound group-0 and move to next group-0: h-[00-12,14]\0... -> h-[00-12,14],... */
+        *(gpattern[0] + gmatch[0][0].rm_eo) = tmp[0];
+        gpattern[0] += gmatch[0][0].rm_eo;
+    }
+
+    /* clean up match patterns */
+    regfree(&gmatch_old[0]);
+    regfree(&gmatch_new[0]);
+    regfree(&gmatch_old[1]);
+    regfree(&gmatch_new[1]);
+    regfree(&rmatch_old);
+    regfree(&rmatch_new);
+    regfree(&ematch_old);
+    regfree(&ematch_new);
+
+    /* free local nodelist */
+    MPL_free(string);
```

Comment:
This should use the old `HYDU_FREE` macro. The MPL updates aren't part of the 3.2.x series.